### PR TITLE
ARROW-725: [Formats/Java] FixedSizeList message and java implementation

### PR DESCRIPTION
**Pull request by elahrvivaz**
~Currently only added minor type for 2-tuples~

### CODE REVIEW

File: java/vector/src/main/codegen/templates/ComplexCopier.java
Comment (wesm) [inline java/vector/src/main/codegen/templates/ComplexCopier.java]: Is there some way to make the particular size not a special case?

Comment (elahrvivaz) [inline java/vector/src/main/codegen/templates/ComplexCopier.java]: Not that I could figure out... I think this goes back to the problem with the java minor types - I believe @julienledem was planning to remove them eventually.

Comment (wesm) [inline java/vector/src/main/codegen/templates/ComplexCopier.java]: is this partially or fully addressed in #409? Can you take a look?

Comment (elahrvivaz) [inline java/vector/src/main/codegen/templates/ComplexCopier.java]: I think #409 would fix it - I'll wait until that is merged and rebase

**Comment by wesm** [comment]
It's safe to rebase now -- I also renamed FixedWidthBinary to FixedSizeBinary for better conformity with this

**Comment by elahrvivaz** [comment]
thanks, I've rebased

File: java/vector/src/main/codegen/templates/AbstractFieldWriter.java
Comment (julienledem) [inline java/vector/src/main/codegen/templates/AbstractFieldWriter.java]: let's hold off adding the fixed width types to the writer APIs.
One of the goals of those APIs is to be able to ingest arbitrary data (for example JSON)  and infer the schema on the fly (hence the PromotableWriter promoting vectors to union if the type is not homogenous). So it won't work well with fixed width list since nothing prevents having more than one size in this api.

File: java/vector/src/main/codegen/templates/AbstractPromotableFieldWriter.java
Comment (julienledem) [inline java/vector/src/main/codegen/templates/AbstractPromotableFieldWriter.java]: 2 things:
 - this kind of change should be in its own pull request.
 - For performance reason we may not want to do it.

File: java/vector/src/main/codegen/templates/AbstractFieldWriter.java
Comment (elahrvivaz) [inline java/vector/src/main/codegen/templates/AbstractFieldWriter.java]: If we don't add it to the apis, how would we write fixed size lists?

File: java/vector/src/main/codegen/templates/AbstractPromotableFieldWriter.java
Comment (elahrvivaz) [inline java/vector/src/main/codegen/templates/AbstractPromotableFieldWriter.java]: I think the only performance penalty is comparing an ArrowType instance instead of an enum compare - seems pretty minor.
Would you like me to split out into a separate PR?
MinorType is still used extensively in the writer framework - I think we ultimately need to pull it out since it doesn't capture the full arrow type. I noticed that the writers don't really handle decimal types, for that reason.
The other alternative is to create minor types for various fixed size lists - that would work for me, as I really only need 2-tuples at the moment. But it isn't especially flexible and @wesm asked for a more generic implementation.

File: java/vector/src/main/codegen/templates/AbstractFieldWriter.java
Comment (elahrvivaz) [inline java/vector/src/main/codegen/templates/AbstractFieldWriter.java]: By the way, arbitrary data should still work fine and default to regular lists - you have to specifically set up the fixed size list if you want to use it. If you know you have a fixed size list, you can use that optimization, otherwise you can just use regular lists that will work the same with slightly more overhead.

**Comment by elahrvivaz** [comment]
@julienledem do you have any more comments, or what do I need to do to get this approved?

File: java/vector/src/main/codegen/templates/AbstractFieldWriter.java
Comment (julienledem) [inline java/vector/src/main/codegen/templates/AbstractFieldWriter.java]: - you can write to the vector directly through its mutator. The Writer api is a facility to write in a row-oriented fashion to the vectors.
- How does this interact with the PromotableWriter? what happens if we write both fixed lists and variable size lists to the same field?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java]: There's a discrepancy here. We compare ArrowType but then we create the writer based on MinorType which the two types may have the same even if they are not equal (Decimal or FixedWidthList for example).
It looks like if we start writing fixed width lists and then we write a different width we will promote but write the different width lists to the old fixed width vector. Possibly we should fail in this case.

File: java/vector/src/main/codegen/templates/AbstractPromotableFieldWriter.java
Comment (julienledem) [inline java/vector/src/main/codegen/templates/AbstractPromotableFieldWriter.java]: Agreed that we need a more flexible solution. A Minor type for every list size is very inconvenient.
Regarding perf: There are a few places were we convert between minorType and arrowType each way. We also use ArrowType.equals instead of == between MinorTypes in getWriter(ArrowType) which is called in startList() which is called for each list creation. Just saying that we need to actually measure the consequence of this. I'm not trying to make you do more work than necessary. Separating the refactoring from the new type and vector allows to get the later merged while we finalize the former.

File: java/vector/src/main/codegen/templates/UnionListWriter.java
Comment (julienledem) [inline java/vector/src/main/codegen/templates/UnionListWriter.java]: we don't really need to got Through ArrowType nor UnionType to call getWriter(...).
We could have a getFixedWidthList(size) in writer.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java]: Would this work for fixed width lists?

File: java/vector/src/main/java/org/apache/arrow/vector/types/pojo/FieldType.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/types/pojo/FieldType.java]: shouldn't this be done from the type in the vector?

File: java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java
Comment (julienledem) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/impl/PromotableWriter.java]: UnionWriter doesn't work for Decimal it looks like (https://github.com/apache/arrow/blob/master/java/vector/src/main/codegen/templates/UnionWriter.java#L124)
It's fine not to support FixedWidthList in union either. (at least for now)

**Review by julienledem — COMMENTED** [review]
Overall this looks good but we need to figure out the change to getWriter(ArrowType). Possibly we can have more than one method (getFixedWidthListWriter...) to avoid creating objects and calling equals. The UnionWriter can be limited to just arbitrary json writing type use case.

**Comment by elahrvivaz** [comment]
@julienledem I will revert the writer changes and just add some unit tests that show how to use the mutators directly.

**Comment by elahrvivaz** [comment]
reverted - ready for another review

**Comment by wesm** [comment]
@julienledem @elahrvivaz I'll be happy to provide a C++ implementation whenever this goes in, then we can set up integration tests. Let me know

**Review by julienledem — APPROVED** [review]
LGTM.
@jacques-n ?

**Comment by julienledem** [comment]
Thank you @elahrvivaz !

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java
Comment (jacques-n) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: Reordering on purpose?

**Review by jacques-n — COMMENTED** [review]
It looks like there is one small spurious import reorder. Otherwise, lgtm.

File: java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java
Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: that's just how my ide (intellij) sorts imports alphabetically. I'll revert.

Comment (elahrvivaz) [inline java/vector/src/main/java/org/apache/arrow/vector/complex/ListVector.java]: reverted

**Comment by elahrvivaz** [comment]
hold off merging - I need to add reAlloc to FixeSizeListVector now that #534 is merged

**Comment by wesm** [comment]
Sounds good

**Comment by elahrvivaz** [comment]
updated and rebased - should be good now

**Comment by wesm** [comment]
+1. This is great, thanks @elahrvivaz!

**Pull request closed without merging**
